src/libmeasurement_kit/dns/parser_impl.hpp

```diff
@@ -0,0 +1,286 @@
+ *
+ *     0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15
+ *   +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ *   |   |   |   |   |   |   |[0]|[1]|[2]|[3]|[4]|[5]|   |   |   |   |
+ *   +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ *                             ^                      ^
+ *                            abuf               abuf + alen
+ *
+ * The question is when the region [aptr, aptr + off) is contained
+ * within the regione [abuf, abuf + alen). This happens when:
+ *
+ *      aptr >= abuf and aptr + off <= abuf + alen
+ *
+ * It follows naturally that `max(off) = alen`. So we can write:
+ *
+ *      aptr >= abuf and off <= alen and aptr + off <= abuf + alen
+ *
+ * Still `aptr + off` can overflow (e.g. with `aptr = 11` and
+ * `off = 6`); to prevent that we can write:
+ *
+ *      aptr >= abuf and off <= alen and aptr - abuf <= alen - off
+ *
+ * Where both the left and the right side of the last comparison are
+ * guaranteed to be always positive by the previous checks.
+ *
+ * The cast is necessary because otherwise we would be comparing integers
+ * of potentially different signedness. And also the cast is valid because
+ * we have excluded cases where the difference is negative.
+ */
+static inline bool valid_mem_region(const unsigned char *aptr, size_t off,
+                                    const unsigned char *abuf, size_t alen) {
+    // Implementation note: checks reorganized to see through coverage
+    // that we're taking all the possible branches
+    if (not (aptr >= abuf)) {
+        return false;
+    }
+    if (not (off <= alen)) {
+        return false;
+    }
+    using ull = unsigned long long;
+    if (not ((ull)(aptr - abuf) <= (ull)(alen - off))) {
+        return false;
+    }
+    return true;
+}
+
+#define VALIDATE_APTR(aptr, abuf, alen)                                        \
+    assert(valid_mem_region(aptr, 0, abuf, alen))
+
+/**
+ * Parse the initial header of a packet.
+ * @param aptr Current pointer within packet.
+ * @param abuf Start of the packet.
+ * @param alen Size of the packet.
+ * @param message Message where to write header fields.
+ * @param logger Logger to use.
+ * @return Updated aptr on success, error on failure.
+ */
+static inline ErrorOr<const unsigned char *>
+parse_header(const unsigned char *aptr, const unsigned char *abuf, size_t alen,
+             Var<Message> message, Var<Logger> logger) {
+    if (!valid_mem_region(aptr, NS_HFIXEDSZ, abuf, alen)) {
+        logger->warn("dns: packet too short: no space for header");
+        return NoSpaceForHeaderError();
+    }
+    message->qid = DNS_HEADER_QID(aptr);
+    message->qr = DNS_HEADER_QR(aptr);
+    message->opcode = DNS_HEADER_OPCODE(aptr);
+    message->aa = DNS_HEADER_AA(aptr);
+    message->tc = DNS_HEADER_TC(aptr);
+    message->rd = DNS_HEADER_RD(aptr);
+    message->ra = DNS_HEADER_RA(aptr);
+    message->z = DNS_HEADER_Z(aptr);
```

Comment:
TODO: mention here that we're ignoring some fields because c-ares apparently does the same